### ISSUE TITLE
Runbook URL adding is too greedy.

### ISSUE THREAD

**Issue by beorn7** [kubernetes-monitoring/kubernetes-mixin#312]
https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/add-runbook-links.libsonnet is a cool way of adding a runbook URL to all alerts in a generic way.

The problem is that this will act on _all_ alerts using the same pattern. If you use other mixins that also define alerts in this pattern, they will (by default) all get a runbook URL pointing to https://github.com/kubernetes-monitoring/kubernetes-mixin/tree/master/runbook.md . Of course, you could set a `runbookURLPattern` that works for all your alerts. But imagine other mixins with alerts would also try to set a generic runbook URL in a similar way to https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/add-runbook-links.libsonnet#L16-L22 . It would probably be fine if they all did exactly the same. But if they did different things, then they will stomp on each others work.

Either we need to make this more specific, so it only sets the runbook URL on alerts from this mixin (and then probably set a more specific config variable than `runbookURLPattern`), or we make this even more generic and make it a "standardized" parts of mixins with alerts.

**Comment by brancz** [comment]
> we make this even more generic and make it a "standardized" parts of mixins with alerts.

We had some discussions around this and I'm very happy to hear suggestions and move this forward if we can come up with something useful.

> Either we need to make this more specific, so it only sets the runbook URL on alerts from this mixi

This touches an area that I've been meaning to discuss with you @tomwilkie, @csmarchbanks and @metalmatze. I think after ~1.5 years of mixins I've (we probably all have) identified a couple of shortcomings of our first iteration, and I feel it might be time to revise some of our decisions. In particular, and that's what would help with this point, I think mixins should not operate on what is essentially global state and global configuration. Each mixin should be instantiated individually and have it's own "private" config, whether a user then deduplicates these in a global configuration is their thing but should not impede the user. If that was the case then we would selectively have these links or explicitly decide to do it globally, but not through unintended side-effect. The changes I have in mind wouldn't be huge, but improve things pretty vastly I believe.

Does that make sense? I feel if we all take a day or so we could pretty easily convert all the mixins we have under control to a new of course first discussed design. If that makes sense, I'd say let's write up the revised mixin design doc and take it from there?

**Comment by paulfantom** [comment]
Can I propose a radical idea of embedding `runbook_url` in an alert annotation itself instead of using https://github.com/kubernetes-monitoring/kubernetes-mixin/blob/master/alerts/add-runbook-links.libsonnet? We do have literally 4 alerts for which we have some meaningful action items defined in `runbook.md`, everything else there is just spam.

**Comment by beorn7** [comment]
So there are different schools about runbooks.

One school claims that most alerts are "self-explaining" so only a few need a runbook. That would gel well with your idea.

Another school says that every alert should have a runbook link. This school has many sub-denominations. For example the strict one where you essentially want a fixed pattern with which the URL for each runbook would be strictly derived from the alert name, e.g. https://wiki.example.org/runbooks/MyAlertName . That's where the configurable pattern comes from. Other denominations will want to allow different teams to define different patterns. In general, I believe that runbooks often have to be specific to an organization so that the generic runbooks we provide in the mixin are only a starting point. Plus, even for alerts where we don't have a runbook in the mixin, an organization will want to provide a runbook for their specific setup.

Another dimension is what runbooks are actually for. Naively, one might assume a runbook is a fairly static explanation of the alert. I think, however, that the best runbooks are ephemeral and dynamic, at least in part. E.g. they are used to track links to currently known bugs or troubleshooting history. That's why something like a Wiki is a good choice for runbooks.

In summary, I think we should make it possible for users of the mixins to adjust runbook links to what they use internally.

**Comment by paulfantom** [comment]
Having "built-in" `runbook_url` annotation doesn't impact the ability to configure it when mixin is consumed. It is a matter of a jsonnet patch that can be specific to an organization.

I don't see much benefit in maintaining a generic patch in the upstream monitoring mixin project so it can be used in various organizations. Such an approach may work for some, but not for all. Especially that such patch should be applied to all consumed mixins, not only to this project.

**Comment by beorn7** [comment]
Any solution that allows a straightforward way to customize the runbook URLs as described above should be fine. This solution might be exactly the one you described. Even better if we can provide a convenience function to inject a pattern (per mixin or globally… this has to be fleshed out, with the possible result of not providing any convenience function is the easiest and most generally usable).
